src/tangent_types/thunks.jl

```diff
@@ -249,3 +249,121 @@ function Base.show(io::IO, x::InplaceableThunk)
+
+Does that mean we should use `BroadcastThunk` even for expensive operations,
+and rely on downstreadm rules to use it just once? Or would a better policy be
+to only use this for cheap operations, and downstreadm rules may then dispatch
+on `::BroadcastThunk` & fuse it into two different places?
+"""
+struct BroadcastThunk{T, B<:Broadcast.Broadcasted} <: AbstractThunk
+    bc::B
+end
+function BroadcastThunk(bc::Broadcast.Broadcasted)
+    T = Base.@default_eltype(bc)
+    return if T <: Number  # applicable(zero, T)  # will SVector work?
+        BroadcastThunk{T, typeof(bc)}(bc)
+    else
+        # We need init=zero(T) for unbroadcast to work.
+        # For things like arrays of arrays, we just don't thunk?
+        # copy(bc)
+        # Or perhaps we make a boring thunk?
+        InplaceableThunk(dx -> dx .+= bc, @thunk copy(bc))
+    end
+end
+function BroadcastThunk(x::AbstractArray{T}) where {T}
+    return if T <: Number  # applicable(zero, T)
+        bc = Broadcast.instantiate(Broadcast.broadcasted(identity, x))
+        BroadcastThunk{T,typeof(bc)}(bc)
+    else
+        x
+    end
+end
+Base.eltype(x::BroadcastThunk{T}) where {T} = T
+
+@inline unthunk(x::BroadcastThunk) = copy(x.bc)
+
+# This is the whole point:
+Base.Broadcast.broadcastable(x::BroadcastThunk) = x.bc
+
+function Base.show(io::IO, x::BroadcastThunk)
+    print(io, "BroadcastThunk{")
+    show(io, eltype(x))
+    print(io, "}(")
+    str = sprint(show, x.bc, context = io)
+    if length(str) < 80
+        printstyled(io, str, color=:light_black)
+    else
+        printstyled(io, str[1:70], "...", color=:light_black)
+    end
+    print(io, ")")
+end
+
+"""
+    @bc_thunk f(a, g(b, c))
+
+This works like `@.` to produce something like `@thunk f.(a, g.(b, c))`.
+Except that instead of a `Thunk`, it's a `BroadcastThunk`.
+"""
+macro bc_thunk(ex)
+  bc = esc(Broadcast.__dot__(ex))
+  :($_lazy_bc.($bc))
+end
+
+function _lazy_bc end
+Broadcast.broadcasted(::typeof(_lazy_bc), x) = _Lazy_BC(x)
+struct _Lazy_BC{T}; bc::T; end
+Broadcast.materialize(x::_Lazy_BC) = BroadcastThunk(Broadcast.instantiate(x.bc))
+
+macro bc_thunk(s::Symbol)
+    error("cannot apply @bc_thunk to one symbol, there is nothing to broadcast!")
+end
+
+# These prove useful for writing rules:
+
+Base.:(-)(x::BroadcastThunk) = @bc_thunk -(x.bc)
+
+for fun in [:conj, :real, :imag, :complex]
+    @eval Base.$fun(x::BroadcastThunk) = BroadcastThunk(Broadcast.instantiate(Broadcast.broadcasted($fun, x.bc)))
+end
+
+Base.sum(x::BroadcastThunk) = sum(x.bc)
+Base.sum(x::BroadcastThunk; dims=:) = sum(x.bc; dims, init=zero(eltype(x)))
+Base.sum(f, x::BroadcastThunk) = sum(f, x.bc)
+
+LinearAlgebra.dot(x::Base.AbstractArrayOrBroadcasted, y::BroadcastThunk{<:Number}) = sum(@bc_thunk conj(x) * y.bc)
+
+"""
+    unthunk_or_bc(dx)
+
+This removes most thunks, but turns a `BroadcastThunk` into a `Broadcasted`.
+For use in rrules which can handle the latter.
+"""
+unthunk_or_bc(x) = unthunk(x)
```

Comment:
maybe `unthunk` should do this on `BroadcastedThunks`?
Rather than a seperate function?
idk maybe not since `Broadcasted` isn't a linear algrebra type.

Reply:
If your rule used to call `unthunk` and get an array, then it may fail if a new kind of thunk doesn't respect that. 

Agree it adds complexity to have two unwrap-this functions.